### PR TITLE
feat(optimizer): add kafka scan operator

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Related issue: https://github.com/risingwavelabs/risingwave/issues/16354
- Add a `LogicalKafkaScan` and `BatchKafkaScan` to avoid coupling too much with the `Source` operator.
- `LogicalKafkaScan` would be converted from `LogiaclSource` at the beginning of batch query optimization.
- `BatchKafkaScan` would still be converted to `SourceNode` by `to_batch_prost_body` to keep the modification minimum.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_kafka_scan.rs [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+use crate::utils::{ColIndexMapping, Condition};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct LogicalKafkaScan {
+    pub base: PlanBase<Logical>,
+    pub core: generic::Source,
+
+    /// Expressions to output. This field presents and will be turned to a `Project` when
+    /// converting to a physical plan, only if there are generated columns.
+    output_exprs: Option<Vec<ExprImpl>>,
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/logical_kafka_scan.rs:48]: Random thought: having `output_exprs` here is a little sign of copy-pasting to me. Maybe we can let `LogicalSource::create` directly return `Project + Source`. (Maybe not in this PR)

File: src/frontend/src/optimizer/plan_node/logical_kafka_scan.rs [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+use crate::optimizer::plan_node::expr_visitable::ExprVisitable;
+use crate::optimizer::plan_node::utils::column_names_pretty;
+use crate::optimizer::plan_node::{
+    BatchKafkaScan, ColumnPruningContext, LogicalSource, PredicatePushdownContext,
+    RewriteStreamContext, ToStreamContext,
+};
+use crate::utils::{ColIndexMapping, Condition};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct LogicalKafkaScan {
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/logical_kafka_scan.rs:42]: Better to add comments here mentioning it's only for batch.

**Review by xxchan — APPROVED** [review]
I like this. Thanks!

File: src/frontend/src/optimizer/plan_node/batch_source.rs
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/batch_source.rs:1]: I'm thinking whether BatchSource can still work if LogicalSource is not converted to LogicalKafkaSource first.

Maybe we can add an assertion like
```
impl ToBatch for LogicalSource {
    fn to_batch(&self) -> Result<PlanRef> {
        if is_kafka {
			unreachable("LogicalSource should be converted to LogicalKafkaScan")
		}
		...
    }
}
```

**Review by BugenZhao — APPROVED** [review]

**Review by tabVersion — APPROVED** [review]

File: src/frontend/src/optimizer/plan_node/logical_kafka_scan.rs (hunk 1 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/logical_kafka_scan.rs:48]: Agree. Let's do it later. `output_exprs` could easily introduce a potential bug, for example, the schema of the `LogicalSource` could be different from its `core` schema which is caught by the equivalent plan checking of the optimizer during the development of this PR.

**Pull request merged**